numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1846,6 +1846,16 @@ array_empty_like(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kwds)
                                             order, dtype, subok);
     Py_DECREF(prototype);
 
+    /* gh-7854: explicitly fill in Py_None for object dtype,
+       otherwise we get weird data corruption
+    */
```

Comment:
Fair point.  I moved it into `PyArray_NewLikeArray`.  As far `empty()`, I only do the filling in `PyArray_NewLikeArray` for when the `dtype` is `object` (the underlying cause of the segfault).

As soon as Appveyor and Travis give green lights, I will add a check in `PyArray_Empty` that will avoid filling a second time of the `dtype` is already `object`.